### PR TITLE
Anchors

### PR DESCRIPTION
**Pull request by nitrocode**
### Description of your changes

Fixes https://github.com/terraform-docs/terraform-docs/issues/408

I have:

- [x] Read and followed terraform-docs' [contribution process].
- [x] All tests pass when I run `make test`.

### How has this code been tested

[contribution process]: https://git.io/JtEzg

Tested this locally and seems to work well with a single test case. Not all test cases pass. Waiting for more discussion.

Is there a way to regenerate the testdata or should I be updating that manually ?

### CODE REVIEW

File: internal/format/markdown_table.go
Comment (khos2ow) [inline internal/format/markdown_table.go:58]: This is going to be a very (very) repetitive change if you want to do it this way. I'd suggest create new template function (like `anchorName` or something) and move the `<a name=....></a>...` into that. So effectively the change will become:

```diff
-				| {{ name .Name }} | {{ tostring .Version | default "n/a" }} |
+				| {{ anchorName .Name }} | {{ tostring .Version | default "n/a" }} |
```

**Review by khos2ow — COMMENTED** [review]
I really like the idea @nitrocode and just wanted to give you a very early feedback.

File: internal/format/markdown_table.go
Comment (nitrocode) [reply internal/format/markdown_table.go:58]: That's a good point. I'll add a new function for this.

Comment (nitrocode) [reply internal/format/markdown_table.go]: Finished. Thanks for the comment!

**Comment by khos2ow** [comment]
Please hold off any major work on testing until #410 is done.

**Comment by nitrocode** [comment]
Ah I just added the `--anchors` switch and pushed it. I will hold off on making additional changes.

File: internal/template/template.go
Comment (khos2ow) [inline internal/template/template.go:64]: This can be simplified as:

```go
			return createAnchor(s, t, settings)
```

File: internal/template/anchor.go
Comment (khos2ow) [inline internal/template/anchor.go:30]: This can be simplified as:

```go
	if settings.ShowAnchors {
		anchorName := fmt.Sprintf("%s_%s", t, s)
		return fmt.Sprintf("<a name=\"%s\"></a> [%s](#%s)", anchorName, s, anchorName)
	}

	return s
```

File: internal/format/testdata/markdown/table-WithRequired.golden
Comment (khos2ow) [inline internal/format/testdata/markdown/table-WithRequired.golden:41]: Instead of modifying `WithRequired` golden file, you can create a new golden file (and its corresponding item in test map for `WithAnchor`. And Please do that for:

- markdown document
- markdown table
- asciidoc document
- asciidoc table

File: cmd/markdown/markdown.go
Comment (khos2ow) [inline cmd/markdown/markdown.go:38]: Please also add the same flag to `asciidoc` command too.

**Review by khos2ow — COMMENTED** [review]
Thank you @nitrocode for starting working on this. This looks ok so far but it needs a bit of changing.

**Comment by nitrocode** [comment]
I'm having issues running test on the `master` branch.

```
$ make test
package github.com/nitrocode/terraform-docs/scripts/docs
	scripts/docs/generate.go:28:2: use of internal package github.com/terraform-docs/terraform-docs/internal/terraform not allowed
```

**Comment by khos2ow** [comment]
I'm wondering why it says `package github.com/nitrocode/terraform-docs/scripts/docs`? Technically you cannot access internal packages from `github.com/terraform-docs/terraform-docs` from `github.com/nitrocode/terraform-docs`.

Are you in GOPATH? If yes, `$GOPATH/src/github.com/terraform-docs/terraform-docs` should be your local cloned folder.

**Comment by nitrocode** [comment]
Thanks. I upgraded golang outside of homebrew and it's working now.

**Comment by nitrocode** [comment]
I was able to fix the commit history by basically deleting my branch and updating the files again... squashing commits before and after an in-between merge does not seem possible.

The asciidoc template doesn't look like it will work currently as anchors are different in asciidoc vs markdown.

Docs
* https://docs.asciidoctor.org/asciidoc/latest/syntax-quick-reference/#links
* https://powerman.name/doc/asciidoc#anchor-1

```asciidoc
[[anchor-1]]
Paragraph or block 1.

anchor:anchor-2[]
Paragraph or block 2.

<<anchor-1>>,
<<anchor-1,First anchor>>,
xref:anchor-2[],
xref:anchor-2[Second anchor].
```

I'll use the above model to update the asciidoc anchors.

**Comment by nitrocode** [comment]
@khos2ow could you review this again please ?

File: cmd/asciidoc/asciidoc.go
Comment (khos2ow) [inline cmd/asciidoc/asciidoc.go:37]: I think singular use of `anchor` is better, throughout the whole PR.

File: internal/template/anchor_test.go
Comment (khos2ow) [inline internal/template/anchor_test.go:32]: Please add same set of tests for `anchors: false` as well.

**Review by khos2ow — COMMENTED** [review]
Couple of minor comments.

**Comment by khos2ow** [comment]
- codecov coverage dropped by 0.29% which can be ignored, but you also can add cases into `templates_test.go#TestBuiltinFunc` to cover newly added functions too.
- that's weird. what does `make docs` do, same error? and why do you have `GOOS=linux`? I thought you were using macos.

**Comment by nitrocode** [comment]
Thanks for reviewing. I'll jump on it today.

Yes, I'm on OSX and the issue was related to the `go1.16` issue again...

Not to add too much scope creep to this PR but it would be nice to use an env var in the `Makefile` like `GOBINARY=go` which can be used in every target then if I need to overwrite it, I can simply do either

```
export GOBINARY=go1.16
make docs
```

or

```
GOBINARY=go1.16 make docs
```

**Comment by khos2ow** [comment]
That's a really good point, I'll add it in another PR! :+1:

**Review by khos2ow — COMMENTED** [review]
One last thing to do, please add `anchor` in:

- `internal/cli/reader.go#L96`
- `docs/reference/config-file.md`

Also please add a bit more information to commit message, as it will be shown in release note.

**Review by khos2ow — APPROVED** [review]
Thank you @nitrocode for all the efforts on this PR! :+1:

**Comment by nitrocode** [comment]
I looked at all the tickets for the `v0.12` project and it seems like this won't be released for some time. Would you be willing to cut a `terraform-docs v0.12-beta1` release like they do with terraform ?

**Comment by khos2ow** [comment]
Yeah sure I'll cut a release soon, but also in the meantime you can use `edge` docker image, which is always HEAD of master. (although that might force a change in your workflow, but at least it's an option to immediately unblock you)

**Comment by nitrocode** [comment]
@khos2ow I'm playing with the new release now and it works great!

One thing. On the release notes, you put @ 9trocode instead of @nitrocode 😅

**Comment by khos2ow** [comment]
Interesting! The list of contributors are being extracted automatically, I should check that script to see why that happened! 😅 

Updated release note, thanks for pointing this out and your contribution!

**Pull request merged**
